### PR TITLE
eth_getpendingtransactions

### PR DESCRIPTION
**Pull request by Bushstar**
Based on the behaviour in the link below.

http://man.hubwiz.com/docset/Ethereum.docset/Contents/Resources/Documents/eth_pendingTransactions.html

### CODE REVIEW

File: lib/ain-grpc/src/rpc.rs [hunk 1]
```diff
@@ -216,6 +220,59 @@ impl MetachainRPCServer for MetachainRPCModule {
             })
     }
 
+    fn get_pending_transaction(&self) -> RpcResult<Vec<EthPendingTransactionInfo>> {
```
Comment (Jouzo) [inline lib/ain-grpc/src/rpc.rs:223]: Can we keep the conversion in the `impls.rs` file so we can keep `rpc.rs` clean 

```
    fn get_pending_transaction(&self) -> RpcResult<Vec<EthPendingTransactionInfo>> {
        get_pool_transactions()
            .map(|txs| {
                txs.into_iter()
                    .flat_map(|tx| EthPendingTransactionInfo::try_from(tx.as_str()))
                    .collect()
            })
            .map_err(|e| Error::Custom(e.to_string()))
    }
 ```
 
 in impls.rs :
 
 ```
 
impl TryFrom<&str> for EthPendingTransactionInfo {
    type Error = TransactionError;

    fn try_from(raw_tx: &str) -> Result<Self, Self::Error> {
        let signed_tx: SignedTx = raw_tx.try_into()?;

        let to = if let Some(signed_to) = signed_tx.to() {
            String::from("0x") + &hex::encode(signed_to.as_fixed_bytes())
        } else {
            String::from("null")
        };

        let input = if signed_tx.data().len() > 0 {
            format!("0x{}", hex::encode(signed_tx.data()))
        } else {
            String::from("0x0")
        };

        let pending_transaction = EthPendingTransactionInfo {
            hash: format!(
                "0x{}",
                hex::encode(signed_tx.transaction.hash().as_fixed_bytes())
            ),
            nonce: format!("0x{}", signed_tx.nonce().to_string()),
            block_hash: String::from(
                "0000000000000000000000000000000000000000000000000000000000000000",
            ),
            block_number: String::from("null"),
            transaction_index: String::from("0x0"),
            from: format!("0x{}", hex::encode(signed_tx.sender.as_fixed_bytes())),
            to,
            value: format!("0x{}", signed_tx.value().to_string()),
            gas: format!("0x{}", signed_tx.gas_limit().to_string()),
            gas_price: format!("0x{}", signed_tx.gas_price().to_string()),
            input,
            v: format!("0x{:x}", signed_tx.v()),
            r: format!("0x{}", hex::encode(signed_tx.r().as_fixed_bytes())),
            s: format!("0x{}", hex::encode(signed_tx.s().as_fixed_bytes())),
        };

        Ok(pending_transaction)
    }
}
```

Comment (Bushstar) [reply lib/ain-grpc/src/rpc.rs:223]: Top tip. Thanks Jeremy.

File: lib/ain-grpc/src/impls.rs [hunk 2]
```diff
@@ -70,3 +70,47 @@ impl TryFrom<TransactionV2> for EthTransactionInfo {
...
+        };
+
+        let pending_transaction = EthPendingTransactionInfo {
+            hash: format!(
+                "0x{}",
+                hex::encode(signed_tx.transaction.hash().as_fixed_bytes())
+            ),
+            nonce: format!("0x{}", signed_tx.nonce()),
+            block_hash: String::from(
+                "0000000000000000000000000000000000000000000000000000000000000000",
```
Comment (shohamc1) [inline lib/ain-grpc/src/impls.rs:99]:
```suggestion
                "0x0000000000000000000000000000000000000000000000000000000000000000",
```

File: lib/ain-grpc/src/impls.rs [hunk 3]
```diff
@@ -70,3 +70,47 @@ impl TryFrom<TransactionV2> for EthTransactionInfo {
...
+            format!("0x{}", hex::encode(signed_tx.data()))
+        } else {
+            String::from("0x0")
+        };
+
+        let pending_transaction = EthPendingTransactionInfo {
+            hash: format!(
+                "0x{}",
+                hex::encode(signed_tx.transaction.hash().as_fixed_bytes())
+            ),
```
Comment (shohamc1) [inline lib/ain-grpc/src/impls.rs:96]: Wonder if we can simplify these. 

```suggestion
            hash: format_hash(signed_tx.transaction.hash()),
```

File: lib/ain-grpc/src/impls.rs [hunk 4]
```diff
@@ -70,3 +70,47 @@ impl TryFrom<TransactionV2> for EthTransactionInfo {
...
+            nonce: format!("0x{}", signed_tx.nonce()),
+            block_hash: String::from(
+                "0000000000000000000000000000000000000000000000000000000000000000",
+            ),
+            block_number: String::from("null"),
+            transaction_index: String::from("0x0"),
+            from: format!("0x{}", hex::encode(signed_tx.sender.as_fixed_bytes())),
+            to,
+            value: format!("0x{}", signed_tx.value()),
```
Comment (shohamc1) [inline lib/ain-grpc/src/impls.rs:105]: `format_number` is available for `U256`.

File: lib/ain-grpc/src/rpc.rs [hunk 5]
```diff
@@ -1,18 +1,12 @@
...
-use crate::codegen::types::{
-    EthGetBlockTransactionCountByHashInput, EthGetBlockTransactionCountByHashResult,
-    EthGetBlockTransactionCountByNumberInput, EthGetBlockTransactionCountByNumberResult,
-    EthGetStorageAtInput, EthGetStorageAtResult, EthTransactionInfo,
-};
+use crate::codegen::types::{EthPendingTransactionInfo, EthTransactionInfo};
 
 use crate::receipt::ReceiptResult;
-use ain_cpp_imports::publish_eth_transaction;
+use ain_cpp_imports::get_pool_transactions;
```
Comment (shohamc1) [inline lib/ain-grpc/src/rpc.rs:6]:
```suggestion
```

File: lib/ain-grpc/src/rpc.rs [hunk 6]
```diff
@@ -228,6 +225,16 @@ impl MetachainRPCServer for MetachainRPCModule {
             })
     }
 
+    fn get_pending_transaction(&self) -> RpcResult<Vec<EthPendingTransactionInfo>> {
+        get_pool_transactions()
```
Comment (shohamc1) [inline lib/ain-grpc/src/rpc.rs:229]: For consistency: https://github.com/DeFiCh/ain/blob/0266fbf5f004534ef25b5441030dfcec5e19f0fa/lib/ain-grpc/src/rpc.rs#L213

```suggestion
        ain_cpp_imports::get_pool_transactions()
```

**Pull request merged**